{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "20182765-0eb2-496a-b4a9-00b2d9a8e34b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ========== Step One -- Don't Modify Anything in This Cell\n",
    "# Import Libraries and Modules\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import subprocess\n",
    "from pathlib import Path\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import NGSanalysis_module as ngs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd329a15-1377-4148-82bb-778ff7eeb6bd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "aeb637a5-09fa-44b1-a7b6-0b071e1e9c36",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ========== Step Three -- Modify Parameters for Your Use Case\n",
    "\n",
    "# Parameters to Change\n",
    "\n",
    "bbmap_bin = '/Users/ekello73/src/bbmap/'\n",
    "samtools_bin = '/Users/ekello73/src/samtools-1.22.1/bin/bin/'\n",
    "\n",
    "# Absolute filepath to your merged paired end reads in a .fastq file format\n",
    "fastq_filepath = \"/Users/ekello73/projects/NGS/datasets/PrashantTagmentation/PD2_merged.fastq\"\n",
    "\n",
    "# Primer sequence to the end of the LE or RE\n",
    "primer = \"CTGAAAAACAACCACCACGACATTAATTTGCGAATAACGACACTAAATTGCGAAAAGCGACATTTAATTTGCGAATGTACA\"\n",
    "# AcCAST LE CTGAAAAACAACCACCACGACATTAATTTGCGAATAACGACACTAAATTGCGAAAAGCGACATTTAATTTGCGAATGTACA\n",
    "# PmcCAST LE CCCTTTATTGGAAGCATAAGCTTGCCGTTGCGGCAAAGTTATGGGTAAAGTCACA\n",
    "\n",
    "# Absolute filepath to CJP003 genome .fasta file\n",
    "referencegenome_filepath = \"/Users/ekello73/projects/NGS/genomes/cJP003_assembly.fasta\"\n",
    "\n",
    "# The first 10 bases in your donor plasmid immediately following the LE or RE -- used to throw away non-enzymatic integrations/recombinations\n",
    "first_ten_in_donor = \"TATATAATGG\"\n",
    "# AcCAST TATATAATGG\n",
    "# PmcCAST ATCCCAATGG\n",
    "\n",
    "# Sequence of the targeting portion of your sgRNA\n",
    "sgRNA_sequence = \"TTTTTGGGCTAGCGATTGAAAACG\"\n",
    "\n",
    "# PmcCAST De novo sgRNAs\n",
    "# Spacer 1 TGAAAGCTGGCGCATGATGACCACCGATA\n",
    "# Spacer 2 TGAACTTTACCCGGTGGTGCATATCGGGGA\n",
    "# Spacer 3 ACTTTATCTGACAGCAGACGTGCACTGGCC\n",
    "# Spacer 4 AGGTTAATGGCGTTTTTGATGTCATTTTCG\n",
    "\n",
    "# AcCAST sgRNAs\n",
    "# Spacer 1 TCTCCATACCCGTTTTTTTGGGCT\n",
    "# Spacer 2 TTTTTGGGCTAGCGATTGAAAACG\n",
    "# Spacer 3 TACACCTATAAAAGAGAGAGCCGT\n",
    "\n",
    "# Length of the CAST's PAM Sequence\n",
    "PAM_len = 3\n",
    "\n",
    "# Is this a naturally occuring CAST system?\n",
    "natural_system = True # Set to false if Novel CAST -- integration site relative to the PAM is different than in natural systems\n",
    "\n",
    "# Filepath for Output files from bbmap -- change to the absolute filepath where you want them exported to, end a dir with \"/\"\n",
    "path_to_output_dir = \"./test/\"\n",
    "\n",
    "# Adds this to the beginning of each output file for simplicity\n",
    "sample_abbreviation = \"PD2\"\n",
    "\n",
    "# Names of files of interest -- change to your liking\n",
    "trimmed_fastq = f\"{sample_abbreviation}_trimmed_reads.fastq\"\n",
    "ForwardInsertions = f\"{sample_abbreviation}_forward_insertions.txt\"\n",
    "ReverseInsertions = f\"{sample_abbreviation}_reverse_insertions.txt\"\n",
    "genome_wide_integrations = f\"{sample_abbreviation}_genome_insertion_histogram.png\"\n",
    "ccdb_integrations = f\"{sample_abbreviation}_CCDB_insertion_histogram.png\"\n",
    "integration_distances = f\"{sample_abbreviation}_Integration_profile.png\"\n",
    "integration_report = f\"{sample_abbreviation}_integration_report.txt\"\n",
    "\n",
    "# Useful variables -- only change if necessary\n",
    "genome_sequence = ngs.fasta_to_string(referencegenome_filepath)\n",
    "ccdb_gene = \"ATGCAGTTTAAGGTTTACACCTATAAAAGAGAGAGCCGTTATCGTCTGTTTGTGGATGTACAGAGTGATATTATTGACACGCCCGGGCGACGGATGGTGATCCCCCTGGCCAGTGCACGTCTGCTGTCAGATAAAGTCTCCCGTGAACTTTACCCGGTGGTGCATATCGGGGATGAAAGCTGGCGCATGATGACCACCGATATGGCCAGTGTGCCGGTCTCCGTTATCGGGGAAGAAGTGGCTGATCTCAGCCACCGCGAAAATGACATCAAAAACGCCATTAACCTGATGTTTTGGGGAATATAA\"\n",
    "ccdb_start_site = genome_sequence.find(ccdb_gene)\n",
    "sgRNA_start_site, PAM_start_site, PAM_end_site, is_rc = ngs.sgRNA_finder(sgRNA_sequence, genome_sequence, PAM_len, natural_system)\n",
    "\n",
    "highlight = [\n",
    "    (sgRNA_start_site, sgRNA_start_site + len(sgRNA_sequence), 'red'), (PAM_start_site, PAM_end_site, 'black')\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fed9e016-223b-40e9-8d3d-8bb129c82a55",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total reads before trimming: 84463\n",
      "Total reads after primer trimming: 11509\n",
      "Total reads after donor removal: 5275\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Max memory cannot be determined.  Attempting to use 3200 MB.\n",
      "If this fails, please add the -Xmx flag (e.g. -Xmx24g) to your command, \n",
      "or run this program qsubbed or from a qlogin session on Genepool, or set ulimit to an appropriate value.\n",
      "java -ea -Xmx3200m -Xms3200m -cp /Users/ekello73/src/bbmap/current/ align2.BBMap build=1 overwrite=true fastareadlen=500 ref=/Users/ekello73/projects/NGS/genomes/cJP003_assembly.fasta\n",
      "Executing align2.BBMap [build=1, overwrite=true, fastareadlen=500, ref=/Users/ekello73/projects/NGS/genomes/cJP003_assembly.fasta]\n",
      "Version 39.11\n",
      "\n",
      "No output file.\n",
      "NOTE:\tDeleting contents of ref/genome/1 because reference is specified and overwrite=true\n",
      "NOTE:\tDeleting contents of ref/index/1 because reference is specified and overwrite=true\n",
      "Writing reference.\n",
      "Executing dna.FastaToChromArrays2 [/Users/ekello73/projects/NGS/genomes/cJP003_assembly.fasta, 1, writeinthread=false, genscaffoldinfo=true, retain, waitforwriting=false, gz=true, maxlen=536670912, writechroms=true, minscaf=1, midpad=300, startpad=8000, stoppad=8000, nodisk=false]\n",
      "\n",
      "Set genScaffoldInfo=true\n",
      "Writing chunk 1\n",
      "Set genome to 1\n",
      "\n",
      "Loaded Reference:\t0.003 seconds.\n",
      "Loading index for chunk 1-1, build 1\n",
      "No index available; generating from reference genome: /Users/ekello73/projects/NGS/TagmentationAnalysis/ref/index/1/chr1_index_k13_c8_b1.block\n",
      "Indexing threads started for block 0-1\n",
      "Indexing threads finished for block 0-1\n",
      "Generated Index:\t0.924 seconds.\n",
      "No reads to process; quitting.\n",
      "\n",
      "Total time:     \t1.940 seconds.\n",
      "Max memory cannot be determined.  Attempting to use 3200 MB.\n",
      "If this fails, please add the -Xmx flag (e.g. -Xmx24g) to your command, \n",
      "or run this program qsubbed or from a qlogin session on Genepool, or set ulimit to an appropriate value.\n",
      "java -ea -Xmx3200m -Xms3200m -cp /Users/ekello73/src/bbmap/current/ align2.BBMap build=1 overwrite=true fastareadlen=500 in=test/trimmed.fastq outm=test/trimmed.bam minid=0.9 ambig=random\n",
      "Executing align2.BBMap [build=1, overwrite=true, fastareadlen=500, in=test/trimmed.fastq, outm=test/trimmed.bam, minid=0.9, ambig=random]\n",
      "Version 39.11\n",
      "\n",
      "Choosing a site randomly for ambiguous mappings.\n",
      "Set MINIMUM_ALIGNMENT_SCORE_RATIO to 0.816\n",
      "Set genome to 1\n",
      "\n",
      "Loaded Reference:\t0.029 seconds.\n",
      "Loading index for chunk 1-1, build 1\n",
      "Generated Index:\t0.375 seconds.\n",
      "Analyzed Index:   \t1.136 seconds.\n",
      "Could not find samtools.\n",
      "Started output stream:\t0.035 seconds.\n",
      "Cleared Memory:    \t0.111 seconds.\n",
      "Processing reads in single-ended mode.\n",
      "Started read stream.\n",
      "Started 8 mapping threads.\n",
      "Detecting finished threads: 0, 1, 2, 3, 4, 5, 6, 7\n",
      "\n",
      "   ------------------   Results   ------------------   \n",
      "\n",
      "Genome:                \t1\n",
      "Key Length:            \t13\n",
      "Max Indel:             \t16000\n",
      "Minimum Score Ratio:  \t0.8162\n",
      "Mapping Mode:         \tnormal\n",
      "Reads Used:           \t5275\t(258084 bases)\n",
      "\n",
      "Mapping:          \t0.284 seconds.\n",
      "Reads/sec:       \t18544.08\n",
      "kBases/sec:      \t907.29\n",
      "\n",
      "\n",
      "Read 1 data:      \tpct reads\tnum reads \tpct bases\t   num bases\n",
      "\n",
      "mapped:          \t 61.8957% \t     3265 \t 89.6328% \t      231328\n",
      "unambiguous:     \t 61.1754% \t     3227 \t 89.0326% \t      229779\n",
      "ambiguous:       \t  0.7204% \t       38 \t  0.6002% \t        1549\n",
      "low-Q discards:  \t  0.0000% \t        0 \t  0.0000% \t           0\n",
      "\n",
      "perfect best site:\t 60.7014% \t     3202 \t 86.8047% \t      224029\n",
      "semiperfect site:\t 60.7014% \t     3202 \t 86.8047% \t      224029\n",
      "\n",
      "Match Rate:      \t      NA \t       NA \t 99.9615% \t      231241\n",
      "Error Rate:      \t  1.7231% \t       57 \t  0.0359% \t          83\n",
      "Sub Rate:        \t  1.6626% \t       55 \t  0.0350% \t          81\n",
      "Del Rate:        \t  0.0605% \t        2 \t  0.0009% \t           2\n",
      "Ins Rate:        \t  0.0000% \t        0 \t  0.0000% \t           0\n",
      "N Rate:          \t  0.1814% \t        6 \t  0.0026% \t           6\n",
      "\n",
      "Total time:     \t2.124 seconds.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Integration report written to '/Users/ekello73/projects/NGS/TagmentationAnalysis/test/PD2_integration_report.txt'\n"
     ]
    }
   ],
   "source": [
    "bbmap_bin = '/Users/ekello73/src/bbmap/'\n",
    "samtools_bin = '/Users/ekello73/src/samtools-1.22.1/bin/bin/'\n",
    "\n",
    "[forward_insertions,reverse_insertions] = ngs.process_reads(fastq_filepath,\n",
    "                                                            primer,\n",
    "                                                            first_ten_in_donor,\n",
    "                                                            referencegenome_filepath,\n",
    "                                                            bbmap_bin,\n",
    "                                                            samtools_bin,\n",
    "                                                            path_to_output_dir)\n",
    "\n",
    "# Save positions\n",
    "Path(f\"{path_to_output_dir}{ForwardInsertions}\").write_text(\"\\n\".join(map(str, forward_insertions)))\n",
    "Path(f\"{path_to_output_dir}{ReverseInsertions}\").write_text(\"\\n\".join(map(str, reverse_insertions)))\n",
    "\n",
    "# Plot Integration profiles in genome and in CCDB gene\n",
    "ngs.plot_insertion_histogram(forward_insertions, reverse_insertions, region_start=0, region_end=len(genome_sequence), title=\"Genome\", highlight_regions=highlight, filename=f\"{path_to_output_dir}{genome_wide_integrations}\")\n",
    "ngs.plot_insertion_histogram(forward_insertions, reverse_insertions, region_start=ccdb_start_site-50, region_end=ccdb_start_site+len(ccdb_gene), title=\"CCDB Gene\", highlight_regions=highlight, filename=f\"{path_to_output_dir}{ccdb_integrations}\")\n",
    "\n",
    "# Plot Integration Profiles\n",
    "modified_forward_insertions = ngs.integration_finder(forward_insertions, ccdb_gene, ccdb_start_site, PAM_start_site, PAM_end_site, natural_system, is_rc)\n",
    "\n",
    "modified_reverse_insertions = ngs.integration_finder(reverse_insertions, ccdb_gene, ccdb_start_site, PAM_start_site, PAM_end_site, natural_system, is_rc)\n",
    "\n",
    "combined_integrations = modified_forward_insertions + modified_reverse_insertions\n",
    "\n",
    "ngs.plot_integration_profile(combined_integrations, filename=f\"{path_to_output_dir}{integration_distances}\")\n",
    "\n",
    "# --- Calculate integration stats ---\n",
    "fwd_total = len(forward_insertions)\n",
    "rev_total = len(reverse_insertions)\n",
    "fwd_on_target = len(modified_forward_insertions)\n",
    "rev_on_target = len(modified_reverse_insertions)\n",
    "total = fwd_total + rev_total\n",
    "on_target_total = fwd_on_target + rev_on_target\n",
    "\n",
    "fwd_pct = 100 * fwd_on_target / fwd_total if fwd_total else 0\n",
    "rev_pct = 100 * rev_on_target / rev_total if rev_total else 0\n",
    "total_pct = 100 * on_target_total / total if total else 0\n",
    "\n",
    "all_lengths = modified_forward_insertions + modified_reverse_insertions\n",
    "mean_len = np.mean(all_lengths) if all_lengths else 0\n",
    "median_len = np.median(all_lengths) if all_lengths else 0\n",
    "std_len = np.std(all_lengths) if all_lengths else 0\n",
    "\n",
    "# --- Format and write report ---\n",
    "report_text = f\"\"\"Integration Analysis Report\n",
    "==================================\n",
    "Total insertions:\n",
    "  Forward strand: {fwd_total}\n",
    "  Reverse strand: {rev_total}\n",
    "  Combined total: {total}\n",
    "\n",
    "On-target insertions (within CCDB region):\n",
    "  Forward strand: {fwd_on_target} ({fwd_pct:.2f}%)\n",
    "  Reverse strand: {rev_on_target} ({rev_pct:.2f}%)\n",
    "  Combined on-target: {on_target_total} ({total_pct:.2f}%)\n",
    "\n",
    "Integration length statistics (bp):\n",
    "  Mean length:   {mean_len:.2f}\n",
    "  Median length: {median_len:.2f}\n",
    "  Std deviation: {std_len:.2f}\n",
    "\"\"\"\n",
    "\n",
    "# Create output directory if needed\n",
    "output_report_path = Path(f\"{path_to_output_dir}{integration_report}\")\n",
    "output_report_path.parent.mkdir(parents=True, exist_ok=True)\n",
    "output_report_path.write_text(report_text)\n",
    "\n",
    "print(f\"Integration report written to '{output_report_path.resolve()}'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86d212ad-f4a7-4baf-aa38-200f223d98cc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
