{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "20182765-0eb2-496a-b4a9-00b2d9a8e34b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ========== Step One -- Don't Modify Anything in This Cell\n",
    "# Import Libraries and Modules\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import subprocess\n",
    "from pathlib import Path\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "dd329a15-1377-4148-82bb-778ff7eeb6bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ========== Step Two -- Don't Modify Anything in This Cell\n",
    "# Define Functions\n",
    "def fasta_to_string(fasta_path):\n",
    "    return str(next(SeqIO.parse(fasta_path, \"fasta\")).seq)\n",
    "\n",
    "def trim_reads_by_primer(fastq_filepath, primer):\n",
    "    total_reads = []\n",
    "    \n",
    "    trimmed_reads = []\n",
    "\n",
    "    primer = primer.upper()\n",
    "\n",
    "    primer_len = len(primer)\n",
    "    for rec in SeqIO.parse(fastq_filepath, \"fastq\"):\n",
    "        seq = str(rec.seq)\n",
    "        total_reads.append(seq)\n",
    "        primer_idx = seq.find(primer)\n",
    "        if primer_idx != -1:\n",
    "            trimmed_seq = seq[primer_idx + primer_len:]\n",
    "            trimmed_qual = rec.letter_annotations[\"phred_quality\"][primer_idx + primer_len:]\n",
    "            trimmed_reads.append((rec.id, trimmed_seq, trimmed_qual))\n",
    "    \n",
    "    if len(trimmed_reads) > 0:\n",
    "        print(f\"Total reads before trimming: {len(total_reads)}\")\n",
    "        return trimmed_reads\n",
    "\n",
    "    else:\n",
    "        print(\"Could not trim reads -- double check your primer sequence and offset.\")\n",
    "\n",
    "def write_trimmed_fastq(trimmed_reads, output_path):\n",
    "    with open(output_path, \"w\") as f:\n",
    "        for read_id, seq, qual in trimmed_reads:\n",
    "            qual_str = ''.join([chr(q + 33) for q in qual])\n",
    "            f.write(f\"@{read_id}\\n{seq}\\n+\\n{qual_str}\\n\")\n",
    "\n",
    "def extract_insertion_sites(bam_file, strand=\"f\"):\n",
    "    result = subprocess.run([\"samtools\", \"view\", bam_file], capture_output=True, text=True, check=True)\n",
    "    positions = []\n",
    "    for line in result.stdout.strip().split(\"\\n\"):\n",
    "        if not line or line.startswith('@'):\n",
    "            continue\n",
    "        fields = line.split(\"\\t\")\n",
    "        pos = int(fields[3])\n",
    "        read_len = len(fields[9])\n",
    "        if strand == \"f\":\n",
    "            positions.append(pos)\n",
    "        else:\n",
    "            positions.append(pos + read_len - 1)\n",
    "    return positions\n",
    "\n",
    "def sgRNA_finder(sgRNA, genome, PAM_len, natural_system):\n",
    "    sgRNA = sgRNA.upper()\n",
    "    genome = genome.upper()\n",
    "    \n",
    "    if natural_system:\n",
    "        if genome.find(sgRNA) != -1:\n",
    "            return genome.find(sgRNA), genome.find(sgRNA), genome.find(sgRNA) - PAM_len, False\n",
    "        elif genome.find(str(Seq(sgRNA).reverse_complement())) != -1:\n",
    "            return genome.find(str(Seq(sgRNA).reverse_complement())), genome.find(str(Seq(sgRNA).reverse_complement()))  + len(sgRNA), genome.find(str(Seq(sgRNA).reverse_complement())) + len(sgRNA) + PAM_len, True\n",
    "        else:\n",
    "            print(\"sgRNA Sequence Not Found in Genome\")\n",
    "    \n",
    "    else:\n",
    "        if genome.find(sgRNA) != -1:\n",
    "            return genome.find(sgRNA), genome.find(sgRNA) + len(sgRNA), genome.find(sgRNA) + len(sgRNA) + PAM_len, False\n",
    "        elif genome.find(str(Seq(sgRNA).reverse_complement())) != -1:\n",
    "            return genome.find(str(Seq(sgRNA).reverse_complement())), genome.find(str(Seq(sgRNA).reverse_complement())), genome.find(str(Seq(sgRNA).reverse_complement())) - PAM_len, True\n",
    "        else:\n",
    "            print(\"sgRNA Sequence Not Found in Genome\")\n",
    "\n",
    "def integration_finder(insertion_idx_list, ccdb_gene, ccdb_start_site, PAM_start_site, PAM_end_site, natural_system, is_rc):\n",
    "    dist_list = []\n",
    "\n",
    "    if natural_system:\n",
    "        if is_rc:\n",
    "            def dist_calc(idx): return (PAM_start_site - idx)\n",
    "\n",
    "        else:\n",
    "            def dist_calc(idx): return (idx - PAM_start_site)\n",
    "    \n",
    "    else:\n",
    "        if is_rc:\n",
    "            def dist_calc(idx): return (PAM_end_site - idx)\n",
    "\n",
    "        else:\n",
    "            def dist_calc(idx): return (idx - PAM_end_site)\n",
    "\n",
    "    for idx in insertion_idx_list:\n",
    "        if ccdb_start_site + len(ccdb_gene) >= idx >= ccdb_start_site:\n",
    "            modified_idx = dist_calc(idx)\n",
    "            dist_list.append(modified_idx)\n",
    "\n",
    "    return dist_list\n",
    "\n",
    "def plot_insertion_histogram(\n",
    "    forward_positions, reverse_positions,\n",
    "    region_start=0, region_end=None,\n",
    "    bins=100, title=\"Insertion Sites\",\n",
    "    highlight_regions=None,\n",
    "    filename=None\n",
    "):\n",
    "\n",
    "    if region_end is None:\n",
    "        region_end = max(forward_positions + reverse_positions)\n",
    "\n",
    "    plt.figure(figsize=(10, 4))\n",
    "    ax = plt.gca()\n",
    "\n",
    "    # Plot histograms\n",
    "    ax.hist(forward_positions, bins=bins, range=(region_start, region_end),\n",
    "            alpha=0.5, label='Top Strand', color='black')\n",
    "    ax.hist(reverse_positions, bins=bins, range=(region_start, region_end),\n",
    "            alpha=0.5, label='Bottom Strand', color='grey')\n",
    "\n",
    "    # Plot horizontal red bars\n",
    "    if highlight_regions:\n",
    "        ymin, _ = ax.get_ylim()\n",
    "        y = ymin - 0.01 * (ax.get_ylim()[1] - ymin)\n",
    "        for region in highlight_regions:\n",
    "            if len(region) == 3:\n",
    "                x_start, x_end, color = region\n",
    "            else:\n",
    "                x_start, x_end = region\n",
    "                color = 'red'\n",
    "            ax.hlines(y=2, xmin=x_start, xmax=x_end, colors=color, linewidth=5)\n",
    "\n",
    "    ax.set_xlabel(\"Genomic Position\")\n",
    "    ax.set_ylabel(\"Insertion Count\")\n",
    "    ax.set_title(title)\n",
    "    ax.set_xlim(region_start, region_end)\n",
    "    ax.legend()\n",
    "    plt.tight_layout()\n",
    "\n",
    "    if filename:\n",
    "        plt.savefig(filename, dpi=300)\n",
    "        plt.close()\n",
    "    else:\n",
    "        plt.show()\n",
    "\n",
    "def plot_integration_profile(data, title=\"Integration Distance From PAM\", xlabel=\"Distance From PAM (bp)\", ylabel=\"Count\", \n",
    "                   xlabel_fontsize=12, ylabel_fontsize=12, title_fontsize=14, \n",
    "                   hist_color='grey', edge_color='black', linewidth=5, dpi=300, xlim=(20,100), filename=None):\n",
    "    \n",
    "    plt.clf()\n",
    "    plt.hist(data, edgecolor=edge_color, color=hist_color, linewidth=linewidth)\n",
    "    plt.title(title, fontsize=title_fontsize)\n",
    "    plt.xlabel(xlabel, fontsize=xlabel_fontsize)\n",
    "    plt.ylabel(ylabel, fontsize=ylabel_fontsize)\n",
    "    \n",
    "    if xlim:\n",
    "        plt.xlim(xlim)\n",
    "    \n",
    "    plt.gcf().set_dpi(dpi)\n",
    "    \n",
    "    plt.tick_params(axis='both', labelsize=10)\n",
    "\n",
    "    if filename:\n",
    "        plt.savefig(filename, dpi=dpi)\n",
    "        plt.close()\n",
    "    else:\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "aeb637a5-09fa-44b1-a7b6-0b071e1e9c36",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ========== Step Three -- Modify Parameters for Your Use Case\n",
    "\n",
    "# Parameters to Change\n",
    "\n",
    "# Absolute filepath to your merged paired end reads in a .fastq file format\n",
    "fastq_filepath = \"/Volumes/groups/kellogrp/home/common/asmiley/NGS/PrashantTagmentation/PD2_merged.fastq\"\n",
    "\n",
    "# Primer sequence to the end of the LE or RE\n",
    "primer = \"CTGAAAAACAACCACCACGACATTAATTTGCGAATAACGACACTAAATTGCGAAAAGCGACATTTAATTTGCGAATGTACA\"\n",
    "# AcCAST LE CTGAAAAACAACCACCACGACATTAATTTGCGAATAACGACACTAAATTGCGAAAAGCGACATTTAATTTGCGAATGTACA\n",
    "# PmcCAST LE CCCTTTATTGGAAGCATAAGCTTGCCGTTGCGGCAAAGTTATGGGTAAAGTCACA\n",
    "\n",
    "# Absolute filepath to CJP003 genome .fasta file\n",
    "referencegenome_filepath = \"/Volumes/groups/kellogrp/home/common/06_ProjectStorage/NovelCAST/ProfilingData/Tagmentation/cJP003_assembly.fa\"\n",
    "\n",
    "# The first 10 bases in your donor plasmid immediately following the LE or RE -- used to throw away non-enzymatic integrations/recombinations\n",
    "first_ten_in_donor = \"TATATAATGG\"\n",
    "# AcCAST TATATAATGG\n",
    "# PmcCAST ATCCCAATGG\n",
    "\n",
    "# Sequence of the targeting portion of your sgRNA\n",
    "sgRNA_sequence = \"TTTTTGGGCTAGCGATTGAAAACG\"\n",
    "\n",
    "# PmcCAST De novo sgRNAs\n",
    "# Spacer 1 TGAAAGCTGGCGCATGATGACCACCGATA\n",
    "# Spacer 2 TGAACTTTACCCGGTGGTGCATATCGGGGA\n",
    "# Spacer 3 ACTTTATCTGACAGCAGACGTGCACTGGCC\n",
    "# Spacer 4 AGGTTAATGGCGTTTTTGATGTCATTTTCG\n",
    "\n",
    "# AcCAST sgRNAs\n",
    "# Spacer 1 TCTCCATACCCGTTTTTTTGGGCT\n",
    "# Spacer 2 TTTTTGGGCTAGCGATTGAAAACG\n",
    "# Spacer 3 TACACCTATAAAAGAGAGAGCCGT\n",
    "\n",
    "# Length of the CAST's PAM Sequence\n",
    "PAM_len = 3\n",
    "\n",
    "# Is this a naturally occuring CAST system?\n",
    "natural_system = True # Set to false if Novel CAST -- integration site relative to the PAM is different than in natural systems\n",
    "\n",
    "# Filepath for Output files from bbmap -- change to the absolute filepath where you want them exported to, end a dir with \"/\"\n",
    "path_to_output_dir = \"/Users/asmiley/Downloads/\"\n",
    "\n",
    "# Adds this to the beginning of each output file for simplicity\n",
    "sample_abbreviation = \"PD2\"\n",
    "\n",
    "# Names of files of interest -- change to your liking\n",
    "trimmed_fastq = f\"{sample_abbreviation}_trimmed_reads.fastq\"\n",
    "ForwardInsertions = f\"{sample_abbreviation}_forward_insertions.txt\"\n",
    "ReverseInsertions = f\"{sample_abbreviation}_reverse_insertions.txt\"\n",
    "genome_wide_integrations = f\"{sample_abbreviation}_genome_insertion_histogram.png\"\n",
    "ccdb_integrations = f\"{sample_abbreviation}_CCDB_insertion_histogram.png\"\n",
    "integration_distances = f\"{sample_abbreviation}_Integration_profile.png\"\n",
    "integration_report = f\"{sample_abbreviation}_integration_report.txt\"\n",
    "\n",
    "# Useful variables -- only change if necessary\n",
    "genome_sequence = fasta_to_string(referencegenome_filepath)\n",
    "ccdb_gene = \"ATGCAGTTTAAGGTTTACACCTATAAAAGAGAGAGCCGTTATCGTCTGTTTGTGGATGTACAGAGTGATATTATTGACACGCCCGGGCGACGGATGGTGATCCCCCTGGCCAGTGCACGTCTGCTGTCAGATAAAGTCTCCCGTGAACTTTACCCGGTGGTGCATATCGGGGATGAAAGCTGGCGCATGATGACCACCGATATGGCCAGTGTGCCGGTCTCCGTTATCGGGGAAGAAGTGGCTGATCTCAGCCACCGCGAAAATGACATCAAAAACGCCATTAACCTGATGTTTTGGGGAATATAA\"\n",
    "ccdb_start_site = genome_sequence.find(ccdb_gene)\n",
    "sgRNA_start_site, PAM_start_site, PAM_end_site, is_rc = sgRNA_finder(sgRNA_sequence, genome_sequence, PAM_len, natural_system)\n",
    "\n",
    "highlight = [\n",
    "    (sgRNA_start_site, sgRNA_start_site + len(sgRNA_sequence), 'red'), (PAM_start_site, PAM_end_site, 'black')\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fed9e016-223b-40e9-8d3d-8bb129c82a55",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ========== Step Four -- Don't Modify Anything in This Cell\n",
    "\n",
    "# 1. Trim reads by primer\n",
    "trimmed_reads = trim_reads_by_primer(fastq_filepath, primer)\n",
    "\n",
    "# 2. Remove donor reads\n",
    "donor_removed_reads = [read for read in trimmed_reads if not read[1].startswith(first_ten_in_donor)]\n",
    "print(f\"Total reads after primer trimming: {len(trimmed_reads)}\")\n",
    "print(f\"Total reads after donor removal: {len(donor_removed_reads)}\")\n",
    "\n",
    "# 3. Write output FASTQ with original qualities preserved\n",
    "trimmed_fastq = Path(f\"{path_to_output_dir}{trimmed_fastq}\")\n",
    "write_trimmed_fastq(donor_removed_reads, trimmed_fastq)\n",
    "\n",
    "# 4. BBMap indexing and alignment\n",
    "subprocess.run([\"bbmap.sh\", f\"ref={referencegenome_filepath}\"], check=True)\n",
    "mapped_bam = trimmed_fastq.with_suffix(\".bam\")\n",
    "subprocess.run([\n",
    "    \"bbmap.sh\", f\"in={trimmed_fastq}\", f\"outm={mapped_bam}\",\n",
    "    \"minid=0.9\", \"ambig=random\"\n",
    "], check=True)\n",
    "\n",
    "# 5. Separate forward/reverse BAM\n",
    "forward_bam = mapped_bam.with_name(mapped_bam.stem + \"_f.bam\")\n",
    "reverse_bam = mapped_bam.with_name(mapped_bam.stem + \"_r.bam\")\n",
    "subprocess.run(f\"samtools view -F 0x10 -h {mapped_bam} | samtools view -bS - > {forward_bam}\", shell=True, check=True)\n",
    "subprocess.run(f\"samtools view -f 0x10 -h {mapped_bam} | samtools view -bS - > {reverse_bam}\", shell=True, check=True)\n",
    "\n",
    "# 6. Proper insertion site extraction\n",
    "forward_insertions = extract_insertion_sites(forward_bam, \"f\")\n",
    "reverse_insertions = extract_insertion_sites(reverse_bam, \"r\")\n",
    "\n",
    "# Save positions\n",
    "Path(f\"{path_to_output_dir}{ForwardInsertions}\").write_text(\"\\n\".join(map(str, forward_insertions)))\n",
    "Path(f\"{path_to_output_dir}{ReverseInsertions}\").write_text(\"\\n\".join(map(str, reverse_insertions)))\n",
    "\n",
    "# Plot Integration profiles in genome and in CCDB gene\n",
    "plot_insertion_histogram(forward_insertions, reverse_insertions, region_start=0, region_end=len(genome_sequence), title=\"Genome\", highlight_regions=highlight, filename=f\"{path_to_output_dir}{genome_wide_integrations}\")\n",
    "plot_insertion_histogram(forward_insertions, reverse_insertions, region_start=ccdb_start_site-50, region_end=ccdb_start_site+len(ccdb_gene), title=\"CCDB Gene\", highlight_regions=highlight, filename=f\"{path_to_output_dir}{ccdb_integrations}\")\n",
    "\n",
    "# Plot Integration Profiles\n",
    "modified_forward_insertions = integration_finder(forward_insertions, ccdb_gene, ccdb_start_site, PAM_start_site, PAM_end_site, natural_system, is_rc)\n",
    "\n",
    "modified_reverse_insertions = integration_finder(reverse_insertions, ccdb_gene, ccdb_start_site, PAM_start_site, PAM_end_site, natural_system, is_rc)\n",
    "\n",
    "combined_integrations = modified_forward_insertions + modified_reverse_insertions\n",
    "\n",
    "plot_integration_profile(combined_integrations, filename=f\"{path_to_output_dir}{integration_distances}\")\n",
    "\n",
    "# --- Calculate integration stats ---\n",
    "fwd_total = len(forward_insertions)\n",
    "rev_total = len(reverse_insertions)\n",
    "fwd_on_target = len(modified_forward_insertions)\n",
    "rev_on_target = len(modified_reverse_insertions)\n",
    "total = fwd_total + rev_total\n",
    "on_target_total = fwd_on_target + rev_on_target\n",
    "\n",
    "fwd_pct = 100 * fwd_on_target / fwd_total if fwd_total else 0\n",
    "rev_pct = 100 * rev_on_target / rev_total if rev_total else 0\n",
    "total_pct = 100 * on_target_total / total if total else 0\n",
    "\n",
    "all_lengths = modified_forward_insertions + modified_reverse_insertions\n",
    "mean_len = np.mean(all_lengths) if all_lengths else 0\n",
    "median_len = np.median(all_lengths) if all_lengths else 0\n",
    "std_len = np.std(all_lengths) if all_lengths else 0\n",
    "\n",
    "# --- Format and write report ---\n",
    "report_text = f\"\"\"Integration Analysis Report\n",
    "==================================\n",
    "Total insertions:\n",
    "  Forward strand: {fwd_total}\n",
    "  Reverse strand: {rev_total}\n",
    "  Combined total: {total}\n",
    "\n",
    "On-target insertions (within CCDB region):\n",
    "  Forward strand: {fwd_on_target} ({fwd_pct:.2f}%)\n",
    "  Reverse strand: {rev_on_target} ({rev_pct:.2f}%)\n",
    "  Combined on-target: {on_target_total} ({total_pct:.2f}%)\n",
    "\n",
    "Integration length statistics (bp):\n",
    "  Mean length:   {mean_len:.2f}\n",
    "  Median length: {median_len:.2f}\n",
    "  Std deviation: {std_len:.2f}\n",
    "\"\"\"\n",
    "\n",
    "# Create output directory if needed\n",
    "output_report_path = Path(f\"{path_to_output_dir}{integration_report}\")\n",
    "output_report_path.parent.mkdir(parents=True, exist_ok=True)\n",
    "output_report_path.write_text(report_text)\n",
    "\n",
    "print(f\"Integration report written to '{output_report_path.resolve()}'\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
